phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,45 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use phd_framework::{
+    disk::DiskSource,
+    test_vm::{DiskBackend, DiskInterface},
+};
+use phd_testcase::*;
+// use tracing::{info, warn};
+
+#[phd_testcase]
+async fn configurable_boot_order(ctx: &Framework) {
+    let mut cfg = ctx.vm_config_builder("configurable_boot_order");
+    cfg.data_disk(
+        "alpine-3-20",
+        DiskSource::Artifact("alpine-3-20"),
+        DiskInterface::Virtio,
+        DiskBackend::File,
+        16,
+    );
+
+    cfg.data_disk(
+        "alpine-3-16",
+        DiskSource::Artifact("alpine-3-16"),
+        DiskInterface::Virtio,
+        DiskBackend::File,
+        24,
+    );
+
+    cfg.boot_order(vec!["alpine-3-20", "alpine-3-16"]);
```

Comment:
and finally this is where i'm *really* seeking input. i was _hoping_ to attach a few disks, or even the same disk a few times, vary the boot device order, and use tools in the guest image to verify which PCI slot the boot device is attached to.

in Alpine and other linuxes i could test this by looking for what disk backs `/`, then check which PCI device `/sys/class/block/<that_device>` is a symlink to. t(he perhaps-more-obvious `/dev/disk/by-path` wouldn't work because the minimal Alpine doesn't have udev.) - at first i thought testing this in minimal Linuxes would just be impossible, checking `/sys/class/` only struck me as i was writing this :)

then testing similar for Windows guests is probably doable with some `diskpart` incantations. so that would be great.

but the real wrinkle is that i'd kind of like to test arbitrary pairs of OSes here - whatever's first in the list should behave the same regardless of the second thing in the list, and more importantly if the first boot entry is not actually bootable, it would be a firmware bug if that made the second item fail to boot when we know it would boot on its own too. now, i realize that's really testing that OVMF works, so maybe it's a bit too overboard. opinions?

Reply:
> now, i realize that's really testing that OVMF works, so maybe it's a bit too overboard. opinions?

I think that sort of test is well within bounds here. PHD (to my mind) is an integration test suite that's focused in particular on making sure the *guest abstraction* works the way that we expect. I think it's reasonable to test guest firmware behavior as part of that. It also fits naturally with the way we generate and consume EDK2 builds, where we would:

- make a change in our fork of the EDK2 repo
- wait for buildomat to build the `OVMF_CODE` images that Propolis consumes
- update references to those images in this repo to refer to the commit hash that buildomat processed in the previous step; this includes
  - the commit hash in the `package` directory that specifies what image should be included in the Propolis zone images that Omicron pulls in
  - the commit hash in the PHD artifacts TOML

The PR for this last change will run Propolis CI with the new firmware bits, which gives us signal about how they work "in context" with the rest of Propolis, booting real guest images, honoring settings from the fwcfg module, etc.

---

> but the real wrinkle is that i'd kind of like to test arbitrary pairs of OSes here

I think this might be tricky, or at least will require us to revisit some unwritten assumptions about how PHD artifacts are supposed to be used. Specifically, PHD assumes that each run selects one guest OS to run against (via the `--default-guest-os-artifact` parameter); that OS has to appear in the artifact TOML/artifact store, but it could be just one of the many possible selections there. My (so far unrealized) idea was to set up periodic test runs that run PHD multiple times, each time selecting a different guest OS of interest from a single artifact TOML that's checked into the repo; then we could see that, e.g., such-and-such change works fine for Debian 11 guests but breaks Windows Server 2019 guests, or works fine with this Linux distro but not that one, and so on.

The main thing is that, beyond the value passed in the `default-guest-os-artifact` switch, I hadn't really envisioned that tests would require any particular artifact name to exist in the store. That's so that folks can easily adapt PHD for their own local testing (e.g. by stashing extra test images on a local disk and referring to them in the artifact TOML) without requiring them to obtain a specific list of prerequisite artifacts beyond the ones they've chosen to test.

We could definitely revisit that strategy, but I think my vision for this particular test was that it would instantiate multiple disks backed by the chosen guest OS artifact, and then try to figure out which one actually got booted from. (If we want a non-bootable disk to test fallback, we can just paste a blank disk into the boot order, though this currently requires the use of a Crucible or in-memory disk.)

Reply:
> We could definitely revisit that strategy, but I think my vision for this particular test was that it would instantiate multiple disks backed by the chosen guest OS artifact, and then try to figure out which one actually got booted from.

yup! that's where i ended up. "figuring out which one actually got booted from" was not immediately obvious, but parsing the EFI variables isn't too bad and gets us to the right spot. i believe none of the tests depend on anything about the guest image (other than these `boot_order.rs` tests all heavily depending on being able to read and write EFI variables)